File: cha2020-06-14/cubic_dice.py
def choose(n, k):
    if 0 <= k <= n:
        ntok = 1
        ktok = 1
        for t in xrange(1, min(k, n - k) + 1):
            ntok *= n
            ktok *= t
            n -= 1
        return ntok / ktok
    else:
        return 0

def numrolls(k, n):
    return sum((-1)**i * choose(n, i) * choose(k-6*i-1, k - 6*i - n) for i in xrange((k-n)/6 + 1))

def rolldice_sum_prob(k, n):
    event = numrolls(k, n)
    return float(event)/(6**n)

File: cha2020-06-21/cubic_dice.R
numrolls <- function(k, n, s){
  i <- seq.int(from = 0, to = (k-n)%/%s )
  sum((-1)**i * choose(n, i) * choose(k - s*i - 1, k - s*i - n))
}

rolldice_sum_prob <- function(sum_, dice_amount, sides = 6){
  numrolls(sum_, dice_amount, sides)/sides**dice_amount
}
